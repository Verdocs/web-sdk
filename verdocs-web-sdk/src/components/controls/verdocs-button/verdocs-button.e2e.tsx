import { newE2EPage,E2EPage } from '@stencil/core/testing';

it('verdocs-button visual diff test', async () => {
  const page: E2EPage = await newE2EPage();
  await page.setContent('<verdocs-button></verdocs-button>');
  await page.compareScreenshot('Comparing visual diff for verdocs-button', {fullPage: false});
});
