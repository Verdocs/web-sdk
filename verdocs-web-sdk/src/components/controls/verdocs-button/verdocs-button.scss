@import '../../../theme.scss';

verdocs-button {
  button {
    font-family: $verdocs-primary-font;
    text-transform: uppercase;
    justify-content: center;
    display: inline-flex;
    align-items: center;
    font-weight: 700;
    cursor: pointer;
    margin: 0;

    .icon {
      margin: 0 12px;

      svg {
        fill: $verdocs-grey-4;
        display: block;
      }
    }

    .label {
      padding: 0 12px;
    }

    // Small-sized button size adjustments
    &.small {
      height: 34px;
      font-size: 13px;

      .icon {
        margin: 0 4px;

        svg {
          margin: 2px 0 0 0;
          height: 20px;
        }
      }

      .label {
        padding: 0 4px;
      }
    }

    // Normal-sized button size adjustments
    &.normal {
      height: 44px;
      font-size: 14px;

      .icon.start {
        margin: 0 4px 0 8px;
      }

      .icon.end {
        margin: 0 8px 0 4px;
      }

      .icon {
        svg {
          margin: 2px 0 0 0;
          height: 22px;
        }
      }

      .label {
        padding: 0 10px;
      }
    }

    // Medium-sized button size adjustments
    &.medium {
      height: 52px;
      font-size: 16px;

      .icon.start {
        margin: 0 0 0 18px;
      }

      .icon.end {
        margin: 0 18px 0 0;
      }

      .icon {
        svg {
          margin: 2px 0 0 0;
          height: 26px;
        }
      }

      .label {
        padding: 0 20px;
      }
    }

    // Large-sized button size adjustments
    &.large {
      height: 60px;
      font-size: 20px;

      .icon.start {
        margin: 0 0 0 24px;
      }

      .icon.end {
        margin: 0 24px 0 0;
      }

      .icon {
        svg {
          margin: 2px 0 0 0;
          height: 32px;
        }
      }

      .label {
        padding: 0 34px;
      }
    }

    // Standard variant is a solid color
    &.standard {
      background: $verdocs-green;
      border-radius: 4px;
      outline: none;
      border: none;
      color: white;

      &:disabled {
        background-color: $disabled-color;
        cursor: default;
      }

      &:active:not([disabled]) {
        filter: brightness(110%);
      }
    }

    // Text variant is just a label
    &.text {
      color: $secondary-color;
      background: transparent;
      font-weight: 500;
      padding: 0 4px;
      outline: none;
      border: none;

      &:disabled {
        color: $disabled-color;
      }

      &:active:not([disabled]) {
        filter: brightness(130%);
      }

      .icon {
        svg {
          fill: $secondary-color;
        }
      }
    }

    // Standard variant is a solid color
    &.outline {
      border: 1px solid $secondary-color;
      color: $secondary-color;
      outline: none;
      background: transparent;
      border-radius: 4px;
      padding: 0 20px;

      &:disabled {
        color: $disabled-color;
        border: 1px solid $disabled-color;
      }

      &:active {
        background: $secondary-color;
        color: white;
      }

      .icon {
        svg {
          fill: $secondary-color;
        }
      }
    }
  }
}
