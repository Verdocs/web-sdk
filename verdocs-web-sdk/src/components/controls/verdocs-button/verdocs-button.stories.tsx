import {html} from 'lit-html';
import {Meta} from '@storybook/web-components';

// Circle with three dots in it
const CircleIcon = `
  <svg xmlns="http://www.w3.org/2000/svg" enable-background="new 0 0 24 24" height="24px" viewBox="0 0 24 24" width="24px" fill="#ffffff">
    <g>
      <rect fill="none" height="24" width="24"/>
    </g>
    <g>
      <g>
        <path d="M12,2C6.48,2,2,6.48,2,12c0,5.52,4.48,10,10,10s10-4.48,10-10C22,6.48,17.52,2,12,2z M12,20c-4.42,0-8-3.58-8-8 c0-4.42,3.58-8,8-8s8,3.58,8,8C20,16.42,16.42,20,12,20z"/>
        <circle cx="7" cy="12" r="1.5"/>
        <circle cx="12" cy="12" r="1.5"/>
        <circle cx="17" cy="12" r="1.5"/>
      </g>
    </g>
  </svg>
`;

const ArrowIcon = `
<svg width="10" height="26" viewBox="0 0 10 26" fill="#ffffff" xmlns="http://www.w3.org/2000/svg">
<g clip-path="url(#clip0_31568_3647)">
<path d="M9.64682 13.431L2.10645 20.9466C1.86503 21.1853 1.47466 21.1853 1.23324 20.9466L0.216218 19.9411C-0.0251976 19.7025 -0.0251976 19.3165 0.216218 19.0778L6.30296 12.9993L0.216218 6.9208C-0.0251976 6.68213 -0.0251976 6.29619 0.216218 6.05752L1.23324 5.05205C1.47466 4.81338 1.86503 4.81338 2.10645 5.05205L9.64682 12.5677C9.88824 12.8064 9.88824 13.1923 9.64682 13.431Z"/>
</g>
<defs>
<clipPath id="clip0_31568_3647">
<rect width="9.86207" height="26"/>
</clipPath>
</defs>
</svg>
`;

export default {
  title: 'Controls/Button',
  component: 'verdocs-button',
  args: {
    label: 'Click Me',
    type: 'button',
    size: 'normal',
    variant: 'standard',
    disabled: false,
  },
  argTypes: {
    startIcon: {defaultValue: false, control: {type: 'boolean'}},
    endIcon: {defaultValue: false, control: {type: 'boolean'}},
    onClick: {
      action: 'click',
      table: {
        disable: true,
      },
    },
  },
} as Meta;

export const Button = ({label, size, variant, disabled, startIcon, endIcon, onClick}) =>
  html`<verdocs-button
    .label=${label}
    .size=${size}
    .variant=${variant}
    .disabled=${disabled}
    .startIcon=${startIcon ? CircleIcon : null}
    .endIcon=${endIcon ? ArrowIcon : null}
    @click=${onClick}
  />`;
