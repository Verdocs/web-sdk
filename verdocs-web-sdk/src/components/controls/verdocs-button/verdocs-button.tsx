import {Component, Prop, h} from '@stencil/core';

/**
 * A simple button
 *
 * ```jsx
 * <verdocs-button label="Click Me" />
 * ```
 */
@Component({
  tag: 'verdocs-button',
  styleUrl: 'verdocs-button.scss',
})
export class VerdocsButton {
  /**
   * The label for the button.
   */
  @Prop() label!: string;

  /**
   * If desired, a prefix icon for the button.
   */
  @Prop() startIcon: string | null = null;

  /**
   * If desired, a suffix icon for the button.
   */
  @Prop() endIcon: string | null = null;

  /**
   * The size (height) of the button.
   */
  @Prop() size: 'small' | 'normal' | 'medium' | 'large' = 'normal';

  /**
   * The type of the button.
   */
  @Prop() type: 'button' | 'submit' | 'reset' = 'button';

  /**
   * The display variant of the button.
   */
  @Prop() variant: 'standard' | 'text' | 'outline' = 'standard';

  /**
   * Whether the button should be disabled.
   */
  @Prop() disabled: boolean = false;

  render() {
    return (
      <button disabled={this.disabled} type={this.type} class={`${this.variant} ${this.size} ${this.type}`}>
        {this.startIcon ? <span class="icon start" innerHTML={this.startIcon} /> : null}
        <span class="button-label">{this.label}</span>
        {this.endIcon ? <span class="icon end" innerHTML={this.endIcon} /> : null}
      </button>
    );
  }
}
