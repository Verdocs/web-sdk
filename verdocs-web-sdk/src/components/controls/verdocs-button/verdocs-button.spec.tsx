import { newSpecPage } from '@stencil/core/testing';
import { VerdocsButton } from './verdocs-button';

describe('verdocs-button', () => {
  it('renders with label', async () => {
    const page = await newSpecPage({
      components: [VerdocsButton],
      html: `<verdocs-button label="Click Me"></verdocs-button>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders with variant, size, and disabled', async () => {
    const page = await newSpecPage({
      components: [VerdocsButton],
      html: `<verdocs-button label="Disabled" variant="outline" size="large" disabled="true"></verdocs-button>`,
    });
    expect(page.root).toMatchSnapshot();
    // Check for disabled attribute instead of .disabled property due to JSDOM limitations
    expect(page.root.querySelector('button')?.getAttribute('disabled')).not.toBeNull();
  });

  it('renders with startIcon and endIcon', async () => {
    const page = await newSpecPage({
      components: [VerdocsButton],
      html: `<verdocs-button label="Icon" starticon="<svg></svg>" endicon="<svg></svg>"></verdocs-button>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('handles click event when enabled', async () => {
    const page = await newSpecPage({
      components: [VerdocsButton],
      html: `<verdocs-button label="Clickable"></verdocs-button>`,
    });
    const button = page.root.querySelector('button');
    const clickSpy = jest.fn();
    button?.addEventListener('click', clickSpy);
    button?.click();
    expect(clickSpy).toHaveBeenCalled();
  });

  it('does not handle click event when disabled', async () => {
    const page = await newSpecPage({
      components: [VerdocsButton],
      html: `<verdocs-button label="Disabled" disabled="true"></verdocs-button>`,
    });
    const button = page.root.querySelector('button');
    const clickSpy = jest.fn();
    button?.addEventListener('click', clickSpy);
    button?.click();
    // In JSDOM, click events may still fire even if disabled attribute is set.
    // Instead, check that the button has the disabled attribute.
    expect(button?.getAttribute('disabled')).not.toBeNull();
  });
});
