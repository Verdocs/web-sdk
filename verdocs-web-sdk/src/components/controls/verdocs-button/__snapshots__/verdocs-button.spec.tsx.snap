// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verdocs-button renders with label 1`] = `
<verdocs-button label="Click Me">
  <button class="button normal standard" type="button">
    <span class="button-label">
      Click Me
    </span>
  </button>
</verdocs-button>
`;

exports[`verdocs-button renders with startIcon and endIcon 1`] = `
<verdocs-button endicon="<svg></svg>" label="Icon" starticon="<svg></svg>">
  <button class="button normal standard" type="button">
    <span class="button-label">
      Icon
    </span>
  </button>
</verdocs-button>
`;

exports[`verdocs-button renders with variant, size, and disabled 1`] = `
<verdocs-button disabled="true" label="Disabled" size="large" variant="outline">
  <button class="button large outline" disabled="" type="button">
    <span class="button-label">
      Disabled
    </span>
  </button>
</verdocs-button>
`;
